src/sdam/monitor.rs

```diff
@@ -67,109 +98,180 @@ impl Monitor {
                     && previous_description
                         .map(|sd| sd.is_available())
                         .unwrap_or(false)
                 {
                     self.handle_error(e).await;
-                    retried = true;
                     self.perform_hello().await
                 } else {
-                    Err(e)
+                    HelloResult::Err(e)
                 }
             }
+            other => other,
         };
 
         match check_result {
-            Ok(reply) => {
-                let server_description =
-                    ServerDescription::new(self.address.clone(), Some(Ok(reply)));
-                self.topology_updater.update(server_description).await
+            HelloResult::Ok(reply) => {
+                let server_description = ServerDescription::new(
+                    self.address.clone(),
+                    Some(Ok(reply)),
+                    self.rtt_monitor_handle.average_rtt(),
+                );
+                self.topology_updater.update(server_description).await;
+                true
+            }
+            HelloResult::Err(e) => {
+                self.handle_error(e).await;
+                false
             }
-            Err(e) => self.handle_error(e).await || retried,
+            HelloResult::Cancelled { .. } => false,
         }
     }
 
-    async fn perform_hello(&mut self) -> Result<HelloReply> {
+    async fn perform_hello(&mut self) -> HelloResult {
         self.emit_event(|| {
             SdamEvent::ServerHeartbeatStarted(ServerHeartbeatStartedEvent {
                 server_address: self.address.clone(),
+                awaited: self.topology_version.is_some(),
             })
         })
         .await;
 
-        let (duration, result) = match self.connection {
-            Some(ref mut conn) => {
-                let command = hello_command(
-                    self.client_options.server_api.as_ref(),
-                    self.client_options.load_balanced,
-                    Some(conn.stream_description()?.hello_ok),
-                );
-                let start = Instant::now();
-                (
-                    start.elapsed(),
-                    run_hello(conn, command, None, &self.client_options.sdam_event_handler).await,
-                )
-            }
-            None => {
-                let mut connection = Connection::connect_monitoring(
-                    self.address.clone(),
-                    self.client_options.connect_timeout,
-                    self.client_options.tls_options(),
-                )
-                .await?;
-
-                let start = Instant::now();
-                let res = self
-                    .handshaker
-                    .handshake(
-                        &mut connection,
-                        None,
-                        &self.client_options.sdam_event_handler,
+        let timeout = if self.connect_timeout().as_millis() == 0 {
```

Comment:
The way the tests are implemented, this timeout is supposed to apply to the entire handshake, including connection establishment, authentication, etc. Previously, we only applied it to establishment.